openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1432 @@
+#include <tbb/enumerable_thread_specific.h>
+
+#include <openvdb/math/Math.h> // for Abs() and isExactlyEqual()
+#include <openvdb/math/Stencils.h> // for GradStencil
+#include <openvdb/util/PagedArray.h>
+#include <openvdb/tree/LeafManager.h>
+#include "LevelSetUtil.h"
+#include "Morphology.h"
+
+#include "Statistics.h"
+#ifdef BENCHMARK_FAST_SWEEPING
+#include <openvdb/util/CpuTimer.h>
+#endif
+
+#include <openvdb/math/Stats.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+/// @brief Converts a scalar fog volume into a signed distance function. Active input voxels
+///        with scalar values above the given isoValue will have NEGATIVE distance
+///        values on output, i.e. they are assumed to be INSIDE the iso-surface.
+///
+/// @return A shared pointer to a signed-distance field defined on the active values
+///         of the input fog volume.
+///
+/// @param fogGrid  Scalar (floating-point) volume from which an
+///                 iso-surface can be defined.
+///
+/// @param isoValue A value which defines a smooth iso-surface that
+///                 intersects active voxels in @a fogGrid.
+///
+/// @param nIter    Number of iterations of the fast sweeping algorithm.
+///                 each performing 2^3 = 8 sweeps.
+///
+/// @note Strictly speaking a fog volume is normalized to the range [0,1] but this
+///       method accepts a scalar volume with an arbitary range, as long as the it
+///       includes the @a isoValue.
+///
+/// @details Topology of output grid is identical to that of the input grid, except
+///          active tiles in the input grid will be converted to active voxels
+///          in the output grid!
+///
+/// @throw  RuntimeError if the iso-surface does not intersected any active
+///         voxels or if it intersects any active tiles in @a fogGrid.
+template<typename GridT>
+typename GridT::Ptr
+fogToSdf(const GridT &fogGrid,
+         typename GridT::ValueType isoValue,
+         int nIter = 1);
+
+/// @brief Given an existing approximate SDF it solves the Eikonal equation for all its
+///        active voxels. Active input voxels with a signed distance value above the
+///        given isoValue will have POSITIVE distance values on output, i.e. they are
+///        assumed to be OUTSIDE the iso-surface.
+///
+/// @return A shared pointer to a signed-distance field defined on the active values
+///         of the input sdf volume.
+///
+/// @param sdfGrid  An approximate signed distance field to the specified iso-surface.
+///
+/// @param isoValue A value which defines a smooth iso-surface that
+///                 intersects active voxels in @a sdfGrid.
+///
+/// @param nIter    Number of iterations of the fast sweeping algorithm.
+///                 each performing 2^3 = 8 sweeps.
+///
+/// @note The only difference between this method and fogToSdf, defined above, is the
+///       convention of the sign of the output distance field.
```

Comment:
With this in mind, why do have fog and sdf as aliases for positive and negative?

To answer my question, I like having these methods for clear legibility.  But I'm wondering if they can't all just invoke a common exposed method that takes an isosurface and direction parameter?  This would also make it easier to expose this on an Extrapolate SOP.

Reply:
What do you mean by a direction parameter? imo, it would be simpler if these were consolidated into a single method which checked the grid metadata to infer its type and handle it accordingly. Something like `sweepToLevelSet`

Reply:
this parameter is only used for "mAboveSign  = isInputSdf ? 1 : -1;". Relying on the grid-class metadata is something we avoid if we can (as in this case where the name of the method clearly states that the input is either an SDF or FOG volume),

Reply:
Agreed, I'd like to use metaclass data at all to change behaviour.  I like the fogToSDF and sdfToFog namings as they are quite clear when reading code.  While "1" and "-1" will make me very confused all the time which way it is working.  But both fancily named ones should call something with an explicit mAboveSign parameter.

Reply:
Not only will this make make naming scheme (what would be the name to replace sdfToSdf and fogToSdf?) but in the past we had refrained from relying (100%) on the meta-class tags. It's obviously a simple change to make but I don't see the advantage over the current more explicit naming